aws/alarms/lambda/notify_slack/notify_slack.ts

```diff
@@ -0,0 +1,159 @@
+import { gunzip } from "zlib";
+import { sendToSlack, sendToOpsGenie } from "./utils.js";
+
+export const handler = async (event: any) => {
+  console.log(JSON.stringify(event));
+  try {
+    if (event.awslogs) {
+      await handleCloudWatchLogEvent(event);
+    } else if (event?.Records?.[0]?.Sns?.Message) {
+      handleSnsEventFromCloudWatchAlarm(event);
+    } else {
+      console.log("No supported event type found.");
+      sendToSlack("Unknown Event", JSON.stringify(event, null, 2), "info");
+    }
+
+    return {
+      statusCode: "SUCCESS",
+    };
+  } catch (error) {
+    console.log("Handler Error: ", error);
+    return {
+      statusCode: "ERROR",
+      error: (error as Error).message,
+    };
+  }
+};
+
+/**
+ * Some log contains metadata information like `2023-07-13T13:48:30.151Z  535741e0-25a9-4c8b-a0a1-25d2b24bf1b4  INFO`
+ * which we want to ignore.
+ *
+ * @returns JSON object or false if the message is not JSON
+ */
+export const safeParseLogIncludingJSON = (message: string) => {
+  try {
+    const jsonPartOfLogMessage = message.slice(message.indexOf("{"));
+    return JSON.parse(jsonPartOfLogMessage);
+  } catch (e) {
+    return false;
+  }
+};
+
+export const getAlarmDescription = (message: string): string => {
+  try {
+    const parsedMessage = JSON.parse(message);
+    return parsedMessage.AlarmDescription ? parsedMessage.AlarmDescription : parsedMessage;
+  } catch (err) {
+    return message;
+  }
+};
+
+/**
+ * @returns the severity level: [error, warning, info, alarm_reset] based on the message
+ */
+export const getSNSMessageSeverity = (message: string): string => {
+  const errorMessages = ["error", "critical"];
+  const warningMessages = ["warning", "failure"];
+  const alarm_ok_status = '"newstatevalue":"ok"'; // This is the string that is returned when the alarm is reset
+
+  message = message.toLowerCase();
+
+  if (message.indexOf("sev1") != -1) return "SEV1";
+
+  for (var errorMessagesItem in errorMessages) {
+    if (
+      message.indexOf(errorMessages[errorMessagesItem]) != -1 &&
+      message.indexOf(alarm_ok_status) == -1 // is not an OK status
+    ) {
+      return "error";
+    } else if (message.indexOf(alarm_ok_status) != -1) {
+      return "alarm_reset";
+    }
+  }
+
+  for (var warningMessagesItem in warningMessages) {
+    if (
+      message.indexOf(warningMessages[warningMessagesItem]) != -1 &&
+      message.indexOf(alarm_ok_status) == -1 // is not an OK status
+    ) {
+      return "warning";
+    } else if (message.indexOf(alarm_ok_status) != -1) {
+      return "alarm_reset";
+    }
+  }
+
+  return "info";
+};
+
+export const handleCloudWatchLogEvent = async (event: any) => {
+  console.log("Received CloudWatch logs event: ", JSON.stringify(event));
+  var payload = Buffer.from(event.awslogs.data as string, "base64");
+
+  return new Promise<void>(function (resolve, reject) {
+    gunzip(payload, function (error, result) {
```

Comment:
Not a big deal either but we could convert that `gunzip` function into a Promise one and then call it before the rest of the logic. That new `extractLog` function could also live outside of `handleCloudWatchLogEvent`.

```
return new Promise<void>(function (resolve, reject) {
  gunzip(payload, function (error, result) {
    if (error) {
      reject(error);
    } else {
      resolve(result);
    }
  });
});
```

Reply:
Of course it would imply that you have to call that new function in a `try/catch` block to make sure you can forward any error resulting from the call of that `extractLog`.